openedx/core/djangoapps/credentials_service/models.py

```diff
@@ -1,3 +1,279 @@
+    """
+    Signatory model to add certificate signatories.
+    """
+    name = models.CharField(max_length=255)
+    title = models.CharField(max_length=255)
+    image = models.ImageField(
+        help_text=_(
+            'Image must be square PNG files. The file size should be under 250KB.'
+        ),
+        upload_to=signatory_assets_path,
+        validators=[validate_image]
+    )
+
+    def __unicode__(self):
+        """Unicode representation. """
+        return 'Signatory {name}, {title}'.format(
+            name=self.name,
+            title=self.title,
+        )
+
+    def save(self, *args, **kwargs):
+        """save the Signatory."""
+        if self.pk is None:
+            temp_image = self.image
+            self.image = None
+            super(Signatory, self).save(*args, **kwargs)
+            self.image = temp_image
+        else:
+            super(Signatory, self).save(*args, **kwargs)
+
+
+# pylint: disable=model-missing-unicode
+class CertificateTemplate(TimeStampedModel):
+    """
+    Certificate Template model to organize content for certificates.
+    """
+
+    name = models.CharField(max_length=255, db_index=True)
+    content = models.TextField(
+        help_text=_('HTML Template content data.')
+    )
+    certificate_type = models.CharField(
+        max_length=255, null=True, blank=True,
+        choices=MODE_CHOICES,
+    )
+    organization_id = models.IntegerField(
+        null=True,
+        blank=True,
+        db_index=True,
+        help_text=_('Organization with which this template is attached.'),
+    )
+
+
+class AbstractCertificate(AbstractCredential):
+    """
+    Abstract Certificate configuration to support multiple type of certificates
+    i.e. Programs, Courses.
+    """
+    signatories = models.ManyToManyField(Signatory)
+    template = models.ForeignKey(CertificateTemplate, null=True, blank=True)
+    title = models.CharField(
+        max_length=255, null=True, blank=True,
+        help_text='Custom certificate title to override default display_name for a course/program.'
+    )
+
+    class Meta(object):
+        abstract = True
+
+
+class CourseCertificate(AbstractCertificate):
+    """
+    Configuration for Course Certificates.
+    """
+    course_id = models.CharField(max_length=255, validators=[validate_course_key])
+    certificate_type = models.CharField(
+        max_length=255,
+        choices=MODE_CHOICES
+    )
+
+    class Meta(object):
+        unique_together = (('course_id', 'certificate_type', 'site'),)
+
+    def __unicode__(self):
+        """Unicode representation. """
+        return 'CourseCertificate {course_id}, {certificate_type}'.format(
+            course_id=self.course_id,
+            certificate_type=self.certificate_type,
+        )
+
+
+class ProgramCertificate(AbstractCertificate):
+    """
+    Configuration for Program Certificates.
+    """
+    program_id = models.PositiveIntegerField(db_index=True, unique=True)
+
```

Comment:
I have tried this locally we can get credentials like this 
`p1 = ProgramCertificate.objects.get(program_id=program_id)`
`p1.user_credentials.all()`

otherwise we have to do 
`p1 = ProgramCertificate.objects.get(program_id=program_id)`
`UserCredential.objects.filter(credential_content_type=ContentType.objects.get_for_model(ProgramCertificate), credential_id=p1.id)`

where `program_id` is extracted for api url.